apps/aecore/lib/aecore/oracle/oracle.ex

```diff
@@ -0,0 +1,272 @@
+defmodule Aecore.Oracle.Oracle do
+  alias Aecore.Structures.OracleRegistrationTxData
+  alias Aecore.Structures.OracleQueryTxData
+  alias Aecore.Structures.OracleResponseTxData
+  alias Aecore.Structures.OracleExtendTxData
+  alias Aecore.Structures.DataTx
+  alias Aecore.Structures.SignedTx
+  alias Aecore.Txs.Pool.Worker, as: Pool
+  alias Aecore.Wallet.Worker, as: Wallet
+  alias Aecore.Chain.Worker, as: Chain
+
+  require Logger
+
+  @type ttl :: %{ttl: non_neg_integer(), type: :relative | :absolute}
+
+  @doc """
+  Registers an oracle with the given requirements for queries and responses,
+  a fee that should be paid by queries and a TTL.
+  """
+  @spec register(map(), map(), non_neg_integer(), non_neg_integer(), ttl()) :: :ok | :error
+  def register(query_format, response_format, query_fee, fee, ttl) do
+    payload = %{
+      query_format: query_format,
+      response_format: response_format,
+      query_fee: query_fee,
+      ttl: ttl
+    }
+
+    tx_data =
+      DataTx.init(
+        OracleRegistrationTxData,
+        payload,
+        Wallet.get_public_key(),
+        fee,
+        Chain.lowest_valid_nonce()
+      )
+
+    {:ok, tx} = SignedTx.sign_tx(tx_data, Wallet.get_private_key())
+    Pool.add_transaction(tx)
+  end
+
+  @doc """
+  Creates a query transaction with the given oracle address, data query
+  and a TTL of the query and response.
+  """
+  @spec query(binary(), any(), non_neg_integer(), non_neg_integer(), ttl(), ttl()) :: :ok | :error
+  def query(oracle_address, query_data, query_fee, fee, query_ttl, response_ttl) do
+    payload = %{
+      oracle_address: oracle_address,
+      query_data: query_data,
+      query_fee: query_fee,
+      query_ttl: query_ttl,
+      response_ttl: response_ttl
+    }
+
+    tx_data =
+      DataTx.init(
+        OracleQueryTxData,
+        payload,
+        Wallet.get_public_key(),
+        fee,
+        Chain.lowest_valid_nonce()
+      )
+
+    {:ok, tx} = SignedTx.sign_tx(tx_data, Wallet.get_private_key())
+    Pool.add_transaction(tx)
+  end
+
+  @doc """
+  Creates an oracle response transaction with the query referenced by its
+  transaction hash and the data of the response.
+  """
+  @spec respond(binary(), any(), non_neg_integer()) :: :ok | :error
```

Comment:
be more specific about `binary()` and `any()`, type redefinitions for your usecase are fine, to give the parameter types semantic names